pkg/runtime/runtime.go

```diff
@@ -1934,3 +1982,128 @@ func (a *DaprRuntime) startReadingFromBindings() error {
 	}
 	return nil
 }
+
+func (a *DaprRuntime) suspendInboundTraffics() error {
+	var merr error
+
+	for name, binding := range a.inputBindings {
+		if closer, ok := binding.(bindings.SuspendableInputBinding); ok {
+			if err := closer.SuspendReading(); err != nil {
+				err = fmt.Errorf("error closing input binding %s: %w", name, err)
+				merr = multierror.Append(merr, err)
+				log.Warn(err)
+			}
+		}
+	}
+
+	for name, pubSub := range a.pubSubs {
+		if subscription, ok := pubSub.(pubsub.SuspendableSub); ok {
+			if err := subscription.SuspendSubscription(); err != nil {
+				err = fmt.Errorf("error closing subscription %s: %w", name, err)
+				merr = multierror.Append(merr, err)
+				log.Warn(err)
+			}
+		}
+	}
+
+	return merr
+}
+
+func (a *DaprRuntime) resumeInboundTraffic() error {
+	var merr error
+
+	for name, binding := range a.inputBindings {
+		if closer, ok := binding.(bindings.SuspendableInputBinding); ok {
+			if err := closer.ResumeReading(); err != nil {
+				err = fmt.Errorf("error resuming reading input binding %s: %w", name, err)
+				merr = multierror.Append(merr, err)
+				log.Warn(err)
+			}
+		}
+	}
+
+	for name, pubSub := range a.pubSubs {
+		if subscription, ok := pubSub.(pubsub.SuspendableSub); ok {
+			if err := subscription.ResumeSubscription(); err != nil {
+				err = fmt.Errorf("error resuming subscription %s: %w", name, err)
+				merr = multierror.Append(merr, err)
+				log.Warn(err)
+			}
+		}
+	}
+
+	return merr
+}
+
+func (a *DaprRuntime) waitUntilAppUnavailable(doneCh chan bool) {
+	for {
+		appAvailable, err := a.ProbeApplicationAvailability()
+		if err != nil {
+			log.Error(err.Error())
+		}
+		if !appAvailable {
+			doneCh <- true
+		}
+		time.Sleep(getAppAvailabilityTime)
+	}
+}
+
+func (a *DaprRuntime) waitUntilAppAvailable(doneCh chan bool) {
+	for {
+		appAvailable, err := a.ProbeApplicationAvailability()
+		if err != nil {
+			log.Error(err.Error())
+		}
+		if appAvailable {
+			doneCh <- true
+		}
+		time.Sleep(getAppAvailabilityTime)
+	}
+}
+
+func (a *DaprRuntime) GracefulShutdown() {
+	err := a.suspendInboundTraffics()
+	if err != nil {
+		log.Error(err.Error())
+	}
+
+	doneChan := make(chan bool)
+	go a.waitUntilAppUnavailable(doneChan)
+	<-doneChan
+
+	a.stopActor()
+	err = a.shutdownComponents()
+	if err != nil {
+		log.Error(err.Error())
+	}
+
+	os.Exit(0)
+}
+
+func (a *DaprRuntime) suspendInboundTrafficUntilAppAvailable() {
```

Comment:
There has to be some timeout here. This can cause the sidecar to hang.

Reply:
But it doesn't make sense to make dapr alive when the app is still down. It checks the status of the app periodically, and this method is always run in a goroutine, so I think it will not block the entire dapr process.

Reply:
If you confirm it's always in the context of a goroutine then this is OK.

Reply:
It will always be called by by a goroutine